src/app/3d/qgs3dmaptoolpointcloudchangeattribute.cpp

```diff
@@ -203,50 +206,128 @@ SelectedPoints Qgs3DMapToolPointCloudChangeAttribute::searchPoints( QgsPointClou
+
   if ( !block )
     return selected;
 
+  const bool filterIsValid = filterExpression.isValid();
+  if ( !filterExpression.prepare( block.get() ) && filterIsValid )
+    return selected;
+
   const QgsVector3D blockScale = block->scale();
   const QgsVector3D blockOffset = block->offset();
 
   const char *ptr = block->data();
   const QgsPointCloudAttributeCollection blockAttributes = block->attributes();
   const std::size_t recordSize = blockAttributes.pointRecordSize();
-  int xOffset = 0, yOffset = 0, zOffset = 0;
+  int xOffset = 0, yOffset = 0, zOffset = 0, categoryAttributeOffset = 0;
   const QgsPointCloudAttribute::DataType xType = blockAttributes.find( QStringLiteral( "X" ), xOffset )->type();
   const QgsPointCloudAttribute::DataType yType = blockAttributes.find( QStringLiteral( "Y" ), yOffset )->type();
   const QgsPointCloudAttribute::DataType zType = blockAttributes.find( QStringLiteral( "Z" ), zOffset )->type();
+  const QgsPointCloudAttribute *categoryAttribute = const_cast<QgsPointCloudAttribute *>( blockAttributes.find( categoryAttributeName, categoryAttributeOffset ) );
+
+  // we should adjust the Z based on the layer's elevation properties scale and offset
+  const double layerZScale = static_cast<const QgsPointCloudLayerElevationProperties *>( layer->elevationProperties() )->zScale();
+  const double layerZOffset = static_cast<const QgsPointCloudLayerElevationProperties *>( layer->elevationProperties() )->zOffset();
+
+  const QgsRectangle searchPolygonBbox = searchPolygon.boundingBox();
+
   for ( int i = 0; i < block->pointCount(); ++i )
   {
+    // ignore filtered out points
+    if ( filterIsValid )
+    {
+      double eval = filterExpression.evaluate( i );
+      if ( !eval || std::isnan( eval ) )
+      {
+        continue;
+      }
+    }
+
+    // if using categorized renderer, point might not be in a visible category
+    if ( categoryAttribute )
+    {
+      const double categoryAttributeValue = categoryAttribute->convertValueToDouble( ptr + i * recordSize + categoryAttributeOffset );
+      if ( !categoryValues.contains( categoryAttributeValue ) )
+        continue;
+    }
+
     // get map coordinates
     double x, y, z;
     QgsPointCloudAttribute::getPointXYZ( ptr, i, recordSize, xOffset, xType, yOffset, yType, zOffset, zType, blockScale, blockOffset, x, y, z );
+    z = z * layerZScale + layerZOffset;
+
+    // check if inside map extent
+    if ( !mapExtent.contains( x, y ) )
+      continue;
 
     // project to screen (map coords -> world coords -> clip coords -> NDC -> screen coords)
-    const QPointF ptScreen = mapToPixel3D.transform( x, y, z );
+    bool isInFrustum;
+    const QPointF ptScreen = mapToPixel3D.transform( x, y, z, &isInFrustum );
+
+    if ( !isInFrustum )
+      continue;
+
+    if ( !searchPolygonBbox.contains( ptScreen.x(), ptScreen.y() ) )
+      continue;
 
     if ( searchPolygon.intersects( QgsGeometry( new QgsPoint( ptScreen.x(), ptScreen.y() ) ) ) )
```

Comment:
You'd get MUCH MUCH better performance by changing this from QgsGeometry::intersects to QgsGeos::intersects and using a prepared geometry here.